### PR TITLE
Add a Makefile

### PR DESCRIPTION
**Pull request by lexidor**
I stole this idea from [hack-json-schema](https://github.com/slackhq/hack-json-schema/blob/master/Makefile).
It encourages hackfmt for less painful collaboration.
Would this be something that we would like to have for hhvm projects?
If yes, I'll duplicate this across all projects.

I have my doubts about automagically pulling down a composer.phar file.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+	./vendor/bin/hacktest tests
+
+lint: hh_autoload
+	./vendor/bin/hhast-lint
+
+autoload:
+	./vendor/bin/hh-autoload
+
+composer-fetch:
+	test -f $(COMPOSER) || mkdir -p `dirname $(COMPOSER)`
```
Comment (fredemmott) [inline Makefile:15]: This needs to be whitespace-safe, both for space-in-homedir and space-in-checkout dir, and isn't

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+	test -f $(COMPOSER) || wget -O $(COMPOSER) https://getcomposer.org/download/1.9.0/composer.phar
```
Comment (fredemmott) [inline Makefile:16]: 1. use `curl` instead; wget is less frequently available (e.g. curl is a standard macos command)
2. don't hardcode the version here. We dont' want to have to change a makefile in every project every time

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+composer-install: composer-fetch
+	php $(COMPOSER) install
+
+composer-update: composer-fetch
+	php $(COMPOSER) update
+
+format:
```
Comment (fredemmott) [inline Makefile:24]: maybe `| xargs grep -lv SignedSource | xargs -n 1` and avoid the format.sh ?

**Review by fredemmott — CHANGES_REQUESTED** [review]
Requesting changes for inline comments, but I"m not really sold on the idea of this.

If it actually becomes widespread, it will need updating in every project - but given it manages the package manager, we can't use the package manager to update it. 

Also, I'm not really seeing much value here, other than providign a specific list of targets; I don't see why `make composer-install` is preferable to `composer install`, or `make lint` is better than `vendor/bin/hhast-lint`; overall, I don't think there's enough gains here to justify the maintenance

**Comment by lexidor** [comment]
@fredemmott I am mostly interested in the `make format`.
Working on oss projects requires a habit of `Save without formatting` in vscode.
If I strip out all the commands except for format, the Makefile itself becomes a bit useless.
If I would make format.sh self contained, would that be better?

**Comment by jjergus** [comment]
I like the general idea, but have some concerns similar to Fred -- we probably don't want to manually create a copy of the same file(s) in every respository. If we ever want to change anything (e.g. add a new command, add a flag to hackfmt or something like that), we'd need to do it manually everywhere. Maybe we can check this into some central repository instead? Or even better, maybe it could be a composer package (`hack-makefile` or something like that) that we just add to all projects' composer.json files?

It might also be possible to integrate this more closely with Composer (https://getcomposer.org/doc/articles/scripts.md) so that we don't have 2 different types of tooling, but I'm not sure how valuable that would be, especially since we're considering replacing Composer with something else eventually. Probably worth looking into after we figure out what the replacement is and for now just do whatever is easiest.

> Also, I'm not really seeing much value here, other than providign a specific list of targets; I don't see why make composer-install is preferable to composer install, or make lint is better than vendor/bin/hhast-lint

I like the ability to specify targets to run multiple other targets (e.g. `make not-sure-what-to-call-this` can run hh_client + lint + tests, etc.)

Some of the commands are also more valuable than others, e.g. `make format` is definitely easier to use than the equivalent pipe of find + grep + hackfmt.

---

Separately -- I think it's worth having a separate discussion about whether we want to enforce hackfmt or not. I don't have a strong opinion but I know some people love/hate the idea. If we decide we want that, the most important thing would probably be to make travis-ci enforce it.

I think there are legitimate cases where a human decides something is more readable with a non-standard formatting, but we can always have a @nofmt directive or something like that for those cases (IDEs could be updated to respect it too). That said all the formatting changes in this PR look good, so that makes me more convinced that it would be OK to enforce.

**Comment by lexidor** [comment]
> Separately -- I think it's worth having a separate discussion about whether we want to enforce hackfmt or not. I don't have a strong opinion but I know some people love/hate the idea. If we decide we want that, the most important thing would probably be to make travis-ci enforce it.

I am 100% onboard for requiring formatting. However, different hhvm versions might treat some files slightly differently. This would mean that you are not able to work effectively if your hhvm version does not match the one that travis is using.

> I think there are legitimate cases where a human decides something is more readable with a non-standard formatting, but we can always have a @nofmt directive or something like that for those cases (IDEs could be updated to respect it too). That said all the formatting changes in this PR look good, so that makes me more convinced that it would be OK to enforce.

hackfmt has such a directive `//hackfmt-ignore`.

**Comment by fredemmott** [comment]
> I am 100% onboard for requiring formatting. However, different hhvm versions might treat some files slightly differently. This would mean that you are not able to work effectively if your hhvm version does not match the one that travis is using.


Yeah, we get away with this at FB by pinning the version of the typechecker and hackfmt in the repo.

**Pull request closed without merging**